### ISSUE TITLE
Support bazel-version specific flags in bazelrc

### ISSUE THREAD

**Issue by rickeylev** [bazelbuild/bazel#24043]
### Description of the feature request:

When a project supports multiple Bazel versions, or is straddling two bazel versions during an upgrade, it can be tough to set flags that only exist in particular Bazel versions. This is because Bazel will (quite reasonably) fail with an error when given an unrecognized flag.

To address this, it would be much easier if the bazelrc syntax supported a way to only use lines for particular bazel versions. A couple proposals for addressing this:

(1) `try-import-version $version_string $config_file` Only import the named config file if the version string matches. e.g. `try-import-version 7.4 %workspace%/bazel7.4.bazelrc`

The nice thing about this is it could, if desired, later be expanded to have simple operators. e.g. `try-import-version >=7 ...`

(2a) `if-supported <FLAGS>` Similar to the `common` psuedo-command, a flag is only used if bazel recognizes it. e.g. `if-supported --some-bazel-flag`

The nice thing about this is it works by feature detection.

(2b) `--undefok <flag>` Similar to how `--undefok` works in absl: if the flag is recognized, parse it; if not, ignore it.  e.g. `common --undefok --someflag`. This is often used so that flags can be set for programs before a flag is supported, which allows easier rollforward/rollback of binaries.

(3) `--enable_bazel_version_specific_config` When enabled, bazel automatically adds e.g. `--config=bazel7.4` (but doesn't fail if no such config exists). This is similar to the `--enable_platform_specific_config` flag.

The nice thing about this is it fits neatly into the existing way rcs can include each other.

### Which category does this issue belong to?

CLI

### What underlying problem are you trying to solve with this feature?

Trying to manage bazelrc file while supporting multiple bazel versions.

### Which operating system are you running Bazel on?

Multiple

### What is the output of `bazel info release`?

7.3.2

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

n/a

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?

```text
n/a
```

### Have you found anything relevant by searching the web?

Some related prior art:
* common allowing unrecognized options: https://github.com/bazelbuild/bazel/pull/20720
* automatic host platform-specific flags: https://github.com/bazelbuild/bazel/issues/22763
* env vars in bazelrc: https://github.com/bazelbuild/bazel/issues/10904

**Comment by meisterT** [comment]
cc @gregestren as this might also be a requirement for project.SCL files.

**Comment by fmeum** [comment]
> Does `always` warn or fail? cc @fmeum 

`always` fails, `common` ignores silently

**Comment by rickeylev** [comment]
Thought of a perk the if-supported idea has: it could also be applicable to
starlark flags. This then gives users a way to handle the same situation
when upgrading rules.

**Comment by meisterT** [comment]
`common` does only ignore silently if the flag exists for other commands

**Comment by adam-azarchs** [comment]
`if-supported` also means you don't have to hunt through release notes to figure out which minor version of bazel added the flag of interest.  It's a bit dangerous, though, because it could hide typos, and you might not remember to switch back to e.g. `common` when no longer necessary, so you might not notice if that flag gets removed in a future version.

The other problem is that it doesn't actually satisfy all use cases, for example if there's a flag that _exists_ in some version of bazel but is broken in some way, or if you want to use some flag that's existed for a while but you only if you can _also_ use some other flag that hasn't.

I like (3) for similarity to existing prior art (`--enable_platform_specific_config`), but the main issue I see with it is how does one define appropriate version ranges for it?  You don't want to have to specify every possible X.Y.Z version, but there are definitely times when some flag is broken prior to some Z release.  You could say that `--enable_bazel_version_specific_config` on bazel version 7.3.2 would be equivalent to `--config=bazel7 --config=bazel7.3 --config=bazel7.3.2 --config=bazel7.3.1 --config=bazel7.3.0` and maybe so on with `7.2.Z`, `7.1.Z`, etc, e.g. like the `^` constraint in npm, but those semantics start getting ugly.  A way to get around that would be something like `--config-if-version=min[,max]=bazel7.3`.

**Comment by rickeylev** [comment]
> how could --enable_bazel_version_config define ranges?

Maybe just keep it simple? Bazel adds 3 configs `--config=X --config=X.Y --config=X.Y.Z` (but it only adds ones where the config name exists).

If you need e.g. 7.0-7.3, then you just have to define 3 configs. A bit verbose, but not too bad. Config inclusion handles a lot of it.

```
build:bazel7.0 --config=bazel7_0_to_3
build:bazel7.1 --config=bazel7_0_to_3
build:bazel7.2 --config=bazel7_0_to_3
build:bazel7.3 --config=bazel7_0_to_3
```

**Comment by adam-azarchs** [comment]
Yes, that would work.  Starts drifting towards "complicated to explain," however.

Another question I'd have is how does this interact with versions of bazel built from source at some untagged version?  e.g. skylib's `versions` library will just return `999999.999999.999999` for development versions, which is fine for `>=` constraints but doesn't really work for this.  I think bazel internally just has an empty string there.  I think it's fair to ignore such things for many kinds of features, but this particular feature does seem like it would be particularly important to support for in-development versions.

**Comment by meteorcloudy** [comment]
Maybe fixing https://github.com/bazelbuild/continuous-integration/issues/2080 could avoid the need of this feature to some extend?

**Comment by rickeylev** [comment]
It would mostly solve it for the projects part of Bazel's CI, yes. I filed this FR for the more general case. It's also easier to do local testing when the extra settings aren't in CI configs.

**Comment by meteorcloudy** [comment]
I guess many people won't like making bazelrc more complicated. But even say we implement this at Bazel@HEAD, we would have to backport this to every minor Bazel versions otherwise `build:bazel7.1` will never be triggered automatically, right.

**Comment by gregestren** [comment]
>  it can be tough to set flags that only exist in particular Bazel versions. 

How much of this request is aggregating the rights flags vs. automatically invoking them?

i.e. you can already define`--config= bazel7.4`.  Devs would have to remember to set it for the version of bazel they're using, which is annoying. But which devs are we talking about and how familiar are they with the version straddling?

**Comment by avdv** [comment]
We also felt the need for this, so we implemented that as part of the bazel wrapper script (see https://github.com/tweag/rules_sh/blob/master/tools/bazel and https://github.com/tweag/rules_sh/blob/master/tools/bazel.bat).

It adds a `--config bazel<version>` flag to each bazel invocation, so it is currently only concerned about the major version but that could be extended.

You also have to make sure to add a config group to your .bazelrc for each Bazel version you want to support. We fall back to a common group when there are no flags to set for some Bazel version:
```
common:bazel7 --noincompatible...
common:bazel6 common
common:bazel5 ...

common:common --noverbose_failures
```
Alternatively, it would also be possible to add the flag only conditionally when the option group is found in .bazelrc.

**Comment by mattyclarkson** [comment]
We do it as so:

```
# Version-specific .bazelrc
try-import ${BAZEL_VERSION}.bazelrc
try-import ${BAZEL_MAJOR_MINOR_PATCH_VERSION}.bazelrc
try-import ${BAZEL_MAJOR_MINOR_VERSION}.bazelrc
try-import ${BAZEL_MAJOR_VERSION}.bazelrc
```

We have some Bash that does environment variable replacement in GitLab CI of the run commands file before Bazel parses it.

The `.bazelrc` already defines `%workspace%` for imports. If it could provide `%bazel.version.major%`, `%bazel.version.minorr%`, `%bazel.version.patch%` replacements then we could do this as so:

```
# Version-specific .bazelrc
try-import %bazel.version%.bazelrc
try-import %bazel.version.major%.%bazel.version.minor%.%bazel.version.patch%.bazelrc
try-import %bazel.version.major%.%bazel.version.minor%.bazelrc
try-import %bazel.version.major%.bazelrc
```

**Comment by meteorcloudy** [comment]
@mattyclarkson That's actually a very clever way to do this, are you willing to send a PR to provide `%bazel.version*%` variables?

**Comment by willstranton** [comment]
Made a [pull request](https://github.com/bazelbuild/bazel/pull/27447) based on mattyclarkson@'s proposal in https://github.com/bazelbuild/bazel/issues/24043#issuecomment-3352747391

**Comment by meteorcloudy** [comment]
Moving the discussion https://github.com/bazelbuild/bazel/pull/27447#issuecomment-3500074253 here, I think if we go with conditional importing, it's better to keep the syntax similar to other ecosystem. 

So proposing (with help of Gemini):

```
# Greater than or equal to: commonly used for feature introduction
try-import-if-bazel-version >= 6.0.0 %workspace%/configs/features.rc

# Strictly greater than: used for post-release targeting
try-import-if-bazel-version > 7.0.0 %workspace%/configs/post_v7.rc

# Strictly less than: used for deprecated flags removed in newer versions
try-import-if-bazel-version < 7.0.0 %workspace%/configs/legacy.rc

# Less than or equal to: caps configuration to a specific version
try-import-if-bazel-version <= 5.4.0 %workspace%/configs/v5_fixes.rc

# Exact match: hotfix for a specific broken release
try-import-if-bazel-version == 6.3.2 %workspace%/configs/hotfix_6.3.2.rc

# Not equal to: excludes broken version from standard config
try-import-if-bazel-version != 7.0.1 %workspace%/configs/standard.rc

# Rolling release: targets a specific pre-release or nightly build
try-import-if-bazel-version >= 8.0.0-pre.20231105.1 %workspace%/configs/nightly.rc

# Release Candidate: targets a specific RC or newer (including the stable release)
try-import-if-bazel-version >= 7.0.0rc1 %workspace%/configs/v7_testing.rc
```

@willstranton @fmeum @lberki @brentleyjones WDYT?

**Comment by meteorcloudy** [comment]
`try-import-if-bazel-version` would be a new command in bazelrc, but I tested, older bazel version won't fail but only output an warning.

**Comment by fmeum** [comment]
The syntax looks pretty reasonable to me. I wouldn't want to encourage making using the full flexibility that this affords, but then again making the syntax uglier and less flexible on purpose seems bad as well.

**Comment by willstranton** [comment]
Thanks, seems good to me.  Only one use case doesn't seem to be handled (please let me know if this is a valid use case or not):

If a flag was introduced in `9.1` and was back ported to `8.5`.  How can we target only the versions with that flag?  If we say `>= 8.5`, then that would incorrectly include `9.0`.

Do we need an additional way to specify minor/patch level targeting ala NPM's tilde ranges: https://docs.npmjs.com/cli/v6/using-npm/semver#tilde-ranges-123-12-1

Following NPM's syntax, you would have `~8.5`, which means `>=8.5` and  `<9.0`.  Mapping back to the meteorcloudy's proposed syntax (for an imaginary `foox` flag) would require these two import statements:

`try-import-if-bazel-version ~ 8.5 %workspace%/configs/newfooxflag.rc`
`try-import-if-bazel-version >= 9.1 %workspace%/configs/newfooxflag.rc`

So this would target `version >= 8.5 && version < 9.0` and `version >= 9.1` (which are the versions that support the new `foox` flag).

---

Listing all the version constraint symbols so far and grouping them in an arbitrary way:

1. `>`
2. `>=`
3. `<`
4. `<=`
5. `==`
6. `!=`
7. `~`

1-4 are the "basic/core" version inequality constraints
5-6 are the "identity" constraints
7 is the "minor/patch level" constraint

**Comment by rickeylev** [comment]
Ohhhhh, I know! Skip the sub-DSL stage; jump to the final "turing complete config file" stage 🙃😂  

```
# File: .bazelrc

import foo.bzl%main

# File foo.bzl
def main():
  lines = []
  if int(native.bazel_version.replace(".", "")) % 2 == 1:
    lines.append("common --whatever")
  return lines
```

Edit: I'm only half-joking. Something my team has is a wrapper `build.py` script, and half its purpose is to setup bazelrc stuff.

In particular, something I like about this is being able to do something like "Change what RBE backend you're using based on your host/user/permissions/etc"

**Comment by willstranton** [comment]
lol @rickeylev - yeah, the more this is discussed, the more I begin to wonder why Bazel wrapper scripts aren't sufficient. 

It was mentioned earlier in the thread in https://github.com/bazelbuild/bazel/issues/24043#issuecomment-2531269172 (i.e. use  `bazelisk` and make use of the `tools/bazel` wrapper). 
 
and @rickeylev it seems you're also using a wrapper script already to get this (and other functionality)? 

Can folks comment, if they haven't already, why you'd prefer it to be a feature baked into `.bazelrc`  vs writing your own custom wrapper script?  I guess the answer is that it's just "yet another thing" that Bazel doesn't provide where the community needs to "roll their own".  Trying to see if there are any additional angles - or if it's just the case that this is a constant pain/irritation for the community (and therefore good to fix/implement even though there exists a workaround).

**Comment by rickeylev** [comment]
Oh yeah, tools/bazel, forgot about that

First, the tools/bazel hook isn't well known. I took a quick attempt at it, got infinite recursion, and nearly gave up; i went back to this thread again, reading through the link earlier, and finding BAZEL_REAL that made this doable.

I have open source projects with RBE and remote caching setup. Who can access what to what level is restricted, though. e.g. untrusted public users can use the read-only cache, certain trusted users can use RBE with read-only cache, others can use RBE with read-write cache. Etc.

I have devs (myself included) who bounce around between projects. The projects use different bazel versions. For a user rc file, I can't easily override every projects tools/bazel. So I have comment/uncomment sections of my user rc depending on what project i'm in.

Finally: using starlark feels natural? It's platform neutral, well defined, no system dependency required (bash, python, coreutils, etc)

**Comment by willstranton** [comment]
https://github.com/bazelbuild/bazel/pull/27675 implements @meteorcloudy's  proposal in https://github.com/bazelbuild/bazel/issues/24043#issuecomment-3502702961

Re: @rickeylev's idea of embedding starlark in `.bazelrc`

I know you're half joking and I suspect the bazel maintainers will frown on this, but I did take a very superficial look into what that might entail.  Currently the parsing of `.bazelrc` happens in the C++ client.  If we want to be able to run starlark, then I think we'd need to have a Starlark parser/runtime in C++. Unfortunately, I only saw [three official implementations of Starlark mentioned](https://github.com/laurentlb/awesome-starlark?tab=readme-ov-file#getting-started): Golang, Rust, and Java.  I guess the C++ code could call out to Golang or the Rust implementation also.  That's as far as I got. 

>  So I have [to] comment/uncomment sections of my user rc depending on what project i'm in.

This seems to be the root issue then?  [Bazel best practices](https://bazel.build/configure/best-practices#bazelrc-file) recommends having a per project user-specific bazelrc:

> `try-import %workspace%/user.bazelrc`

This would then allow you to have project-user-specific configuration and avoid the comment/uncommenting you have to do when jumping between projects.

**Comment by lberki** [comment]
I floated the idea of embedding a Starlark interpreter into the launcher and implementing a new bazelrc syntax with it at BazelCon, but no one seemed to be particularly enthusiastic at the idea (notably, @meisterT and @meteorcloudy ) and when pressed, they started looking at me funny, so I concluded that it's me who is misguided here.

I don't want to take on either Rust or Golang as a dependency. If we wanted to do this, I was hoping that I could persuade a friendly AI to rewrite, say, the Golang implementation to C++ (that one because that's the simplest) and if that works, the only tiny little thing that remains is maintaining the resulting code...

**Comment by meisterT** [comment]
>  the only tiny little thing that remains is maintaining the resulting code...

well, I wouldn't want to maintain a second implementation, especially one written in C++

**Comment by lberki** [comment]
Yeah, which is why I mentally classify this as "seductive but not to be taken lightly".

**Comment by alexeagle** [comment]
> embedding a Starlark interpreter into the launcher

that's exactly what we did with [AXL](https://docs.aspect.build/cli/overview#aspect-extension-language), which is fully Apache 2. @lberki can I give you a tour sometime? :)

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 9.0.0 RC5](https://github.com/bazelbuild/bazel/releases/tag/9.0.0rc5). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=9.0.0rc5. Thanks!

**Comment by fmeum** [comment]
@bazel-io fork 8.7.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 8.7.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/8.7.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=8.7.0rc1`. Thanks!
